### PR TITLE
feat: free tx mining

### PR DESCRIPTION
**Pull request by Jakub123Jakub123**

### CODE REVIEW

File: src/miner.cpp [hunk 1]
```diff
@@ -312,6 +353,13 @@ void BlockAssembler::addPackageTxs(int &nPackagesSelected, int &nDescendantsUpda
     // mempool has a lot of entries.
     const int64_t MAX_CONSECUTIVE_FAILURES = 1000;
     int64_t nConsecutiveFailed = 0;
+    uint64_t nPaidTxSize = 0;
+    auto freeTxMaxSizeInBlock = chainparams.GetConsensus().freeTxMaxSizeInBlock;
+
+    if (freeTxMaxSizeInBlock > (nBlockMaxWeight * .25) ) {
```
Comment (mtwaro) [inline src/miner.cpp:359]: No sense in validating the consensus parameter here. It should be taken as it is.

File: test/functional/feature_staking_free_tx_preference.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                    )
+                )
+
+            if tx in free_tx_list:
+                free_tx_size = free_tx_size + tx_size
+            else:
+                nonfree_tx_size = nonfree_tx_size + tx_size
+
+        free_tx_percentage = free_tx_size / self.blockmaxweight
+        assert free_tx_percentage < .28 and free_tx_percentage > .24 # TODO figure out how to make it more precise
```
Comment (mtwaro) [inline test/functional/feature_staking_free_tx_preference.py:45]: The desired block size should be defined as a constant at the beginning of the test and should correspond to the parameter in chainparams. 
Also, consensus has the max free tx size defined as an absolute value, not as the percentage of the block size. If we use the lowered blockweight we're in fact skipping the consensus validation of the mining algorithm, which is a very important case. I would try testing on a full block size.
Also, if the upper limit has to be as big as .28, the current algorithm is not working properly - it shouldn't pass any block with more than 25% of free tx size.

File: test/functional/feature_staking_free_tx_preference.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+        # load free txs into mempool
+        free_tx_list = []
+        for _ in range(amount_if_tested_free_tx):
+            free_tx_id = self.send_free_tx(dummy_addresses[:5], free_tx_amount, 0, addr2)
+            assert free_tx_id is not None
+            free_tx_list.append(free_tx_id)
+
+        self.assert_block_tx_preference(free_tx_list)
+
```
Comment (mtwaro) [inline test/functional/feature_staking_free_tx_preference.py:98]: I would also check two things:
- if the block is around full size.
- if there are any free-tx leftovers in the mempool

If either of above is false it means we're not testing any preference - we're just putting all of the transactions into a block.

File: test/functional/feature_staking_free_tx_preference.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        ##############################################
+        #                                            #
+        #           Test free and paid tx            #
+        #                                            #
+        ##############################################
+
+        self.generate_inputs_for_free_txs(amount_of_tested_free_tx, addr2)
+
+        self.assert_mempool_leftovers(0)
```
Comment (mtwaro) [inline test/functional/feature_staking_free_tx_preference.py:136]: I assume this is temporary - we want to have mempool leftovers to test if the preference worked correctly

Comment (Jakub123Jakub123) [reply test/functional/feature_staking_free_tx_preference.py:136]: It is to test if we have empty mempool before sending transactions for the test

File: src/miner.cpp [hunk 5]
```diff
@@ -401,6 +424,34 @@ void BlockAssembler::addPackageTxs(int &nPackagesSelected, int &nDescendantsUpda
...
+            else{
+                if (fUsingModified) {
+                    mapModifiedTx.get<ancestor_score>().erase(modit);
+                    failedTx.insert(iter);
+                }
+                continue; // cannot fit more free transactions
+            }
+        }
+        else{
+            nPaidTxWeight += txWieght;
```
[the remark this answers is not in the document]
Comment (Jakub123Jakub123) [reply src/miner.cpp]: fixed

File: src/txmempool.h [hunk 6]
```diff
@@ -290,6 +293,9 @@ class CompareTxMemPoolEntryByAncestorFee
         double f1 = a_mod_fee * b_size;
         double f2 = a_size * b_mod_fee;
 
+        if (a_mod_fee == 0) return false;
```
[the remark this answers is not in the document]
Comment (Jakub123Jakub123) [reply src/txmempool.h]: fixed

File: src/txmempool.h [hunk 7]
```diff
@@ -217,6 +217,9 @@ class CompareTxMemPoolEntryByDescendantScore
...
+        if (b_mod_fee == 0) return true;
```
[the remark this answers is not in the document]
Comment (Jakub123Jakub123) [reply src/txmempool.h]: fixed

File: src/miner.cpp [hunk 8]
```diff
@@ -401,6 +424,34 @@ void BlockAssembler::addPackageTxs(int &nPackagesSelected, int &nDescendantsUpda
...
+        if (fUsingModified) {
+            txWieght = modit->nTxWeight;
+        }
+
+        // Free or Paid
+        if(packageFees == 0){
+            if (
+                (nFreeTxWeight + txWieght) < freeTxMaxSizeInBlock
+                &&
+                ((nFreeTxWeight + txWieght) < ((nBlockMaxWeight*.25)-4000))
```
[the remark this answers is not in the document]
Comment (Jakub123Jakub123) [reply src/miner.cpp]: changed

File: src/miner.cpp [hunk 9]
```diff
@@ -401,6 +423,34 @@ void BlockAssembler::addPackageTxs(int &nPackagesSelected, int &nDescendantsUpda
...
+        if (fUsingModified) {
+            txWeight = modit->nTxWeight;
+        }
+
+        // Free or Paid
+        if(packageFees == 0){
+            if (
+                (nFreeTxWeight + txWeight) < chainparams.GetConsensus().freeTxMaxSizeInBlock
+                &&
+                ((nFreeTxWeight + txWeight) < ((nBlockMaxWeight*.25)-4000))
```
Comment (mtwaro) [inline src/miner.cpp:436]: ?

File: src/miner.cpp [hunk 10]
```diff
@@ -387,7 +418,22 @@ void BlockAssembler::addPackageTxs(int &nPackagesSelected, int &nDescendantsUpda
...
+            *iter,
+            ancestors,
+            nNoLimit,
+            nNoLimit,
+            nNoLimit,
+            nNoLimit,
+            dummy,
+            false,
+            chainparams.GetConsensus().freeTxMaxSizeInBlock,
+            nBlockMaxWeight*.25,
```
Comment (mtwaro) [inline src/miner.cpp:432]: No magic numbers

File: src/miner.cpp [hunk 11]
```diff
@@ -401,6 +447,34 @@ void BlockAssembler::addPackageTxs(int &nPackagesSelected, int &nDescendantsUpda
...
+        if (fUsingModified) {
+            txWeight = modit->nTxWeight;
+        }
+
+        // Free or Paid
+        if(txFee == 0){
+            if (
+                ((nFreeTxWeight + txWeight) < chainparams.GetConsensus().freeTxMaxSizeInBlock)
+                &&
+                ((nFreeTxWeight + txWeight) < ((nBlockMaxWeight*.25)))
```
Comment (mtwaro) [inline src/miner.cpp:460]: Magic number

**Review by mtwaro — APPROVED** [review]

**Pull request merged**